### PR TITLE
Book CRUD

### PR DESCRIPTION
**Pull request by cramallo**
## Summary

- Add CRUD book endpoints.
- Add methods to book service.
- Add handling errors.

## Trello Card

https://trello.com/c/3v0bpU0n/18-crear-las-operaciones-crud-de-book

### CODE REVIEW

File: src/main/java/com/wolox/training/dtos/BookDTO.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
+package com.wolox.training.dtos;
+
+public class BookDTO {
+
+    private int id;
```
Comment (IgnOliveto) [inline src/main/java/com/wolox/training/dtos/BookDTO.java]: Why if the id of the book class is a long the dto is integer? we should maintain uniformity of types between the dtos and the classes they represent.

File: src/main/java/com/wolox/training/services/BookService.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.List;
+
+@Service
+public class BookService {
+    @Autowired
+    BookRepository bookRepository;
+
+    public Book createBook(Book book){
+        if(!checkBookFields(book)){
+            throw new BadRequestException("Invalid book fields");
```
Comment (IgnOliveto) [inline src/main/java/com/wolox/training/services/BookService.java]: We should use a class that contains the error messages as constants, we should not have them literal in the place where the exception is raised

File: src/main/java/com/wolox/training/services/BookService.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        }
+        try {
+            return bookRepository.save(book);
+        }catch(ServerErrorException serverError){
+            throw new ServerErrorException("Internal server error");
```
Comment (IgnOliveto) [inline src/main/java/com/wolox/training/services/BookService.java]: We should use a class that contains the error messages as constants, we should not have them literal in the place where the exception is raised

File: src/main/java/com/wolox/training/services/BookService.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        }
+    }
+
+    public List<Book> getBooks(){
+        try {
+            return bookRepository.findAll();
+        }catch(ServerErrorException serverError){
+            throw new ServerErrorException("Internal server error");
```
Comment (IgnOliveto) [inline src/main/java/com/wolox/training/services/BookService.java]: We should use a class that contains the error messages as constants, we should not have them literal in the place where the exception is raised

File: src/main/java/com/wolox/training/services/BookService.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        }
+    }
+
+    public Book updateBook(int id, Book updatedBook){
+        if(!checkBookFields(updatedBook)){
+            throw new BadRequestException("Invalid book fields");
```
Comment (IgnOliveto) [inline src/main/java/com/wolox/training/services/BookService.java]: We should use a class that contains the error messages as constants, we should not have them literal in the place where the exception is raised

File: src/main/java/com/wolox/training/services/BookService.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    public void deleteBook (int id){
+        try {
+            bookRepository.deleteById(id);
+        }catch(ServerErrorException serverError){
+            throw new ServerErrorException("Internal server error");
+        }
+    }
+
+    private Book findById(int id){
+        return bookRepository.findById(id).orElseThrow(()->new NotFoundException("book not found"));
```
Comment (IgnOliveto) [inline src/main/java/com/wolox/training/services/BookService.java]: We should use a class that contains the error messages as constants, we should not have them literal in the place where the exception is raised

File: src/main/java/com/wolox/training/controllers/BookController.java [hunk 7]
```diff
@@ -1,14 +1,59 @@
...
 import org.springframework.ui.Model;
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.*;
 
-@Controller
+import java.util.List;
+import java.util.stream.Collectors;
+
+@RestController
```
Comment (rodrigofranc) [inline src/main/java/com/wolox/training/controllers/BookController.java:14]: When define the endpoint and you have a prefix for all routes, you can include the annotation @RequestMapping in class definition, like this:

```
@RestController
@RequestMapping("/book")
public class BookController {
```
then for example in the delete method:

```
@DeleteMapping("/{id}")
public void deleteBook(@PathVariable("id") int bookId){
```

File: src/main/java/com/wolox/training/controllers/BookController.java [hunk 8]
```diff
@@ -1,14 +1,59 @@
...
     }
+
+    @GetMapping("/book")
+    public List<BookDTO> getBooks(){
+        List<Book> books = bookService.getBooks();
+        return books.stream().map(book -> convertToDto(book)).collect(Collectors.toList());
+    }
+
+    @PostMapping("/book")
+    public BookDTO createBook(@RequestBody BookDTO bookDto){
```
Comment (rodrigofranc) [inline src/main/java/com/wolox/training/controllers/BookController.java:47]: For the rest standard it is a good practice to include the status code according to the operation. You can do this with annotations in spring. For example: @ResponseStatus(HttpStatus.CREATED)

File: src/main/java/com/wolox/training/services/BookService.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+import com.wolox.training.exceptions.ServerErrorException;
+import com.wolox.training.repositories.BookRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import com.wolox.training.models.Book;
+
+import java.util.List;
+
+@Service
+public class BookService {
```
Comment (IgnOliveto) [inline src/main/java/com/wolox/training/services/BookService.java:14]: I liked that you use services to encapsulate logic that should not be in the controller, but in this method validations are made of the attributes of objects of the book class that should be done with annotation at  model level.

**Review by IgnOliveto — CHANGES_REQUESTED** [review]

File: src/main/java/com/wolox/training/models/Book.java [hunk 10]
```diff
@@ -22,10 +25,11 @@
     private String year;
     @Column(nullable = false)
     private int pages;
-    @Column(nullable = false)
+    @Column(nullable = false,unique=true)
     private String isbn;
 
-    public Book(String genre, String author, String image, String title, String subtitle, String publisher, String year, int pages, String isbn) {
+    public Book(String genre, @NotNull String author,@NotNull String image, @NotNull String title,@NotNull String subtitle,
```
Comment (IgnOliveto) [inline src/main/java/com/wolox/training/models/Book.java]: This validations only work in this constructor, look the validation in this [example](https://www.baeldung.com/javax-validation)

**Review by IgnOliveto — CHANGES_REQUESTED** [review]

**Review by IgnOliveto — APPROVED** [review]
Good Work!

**Review by rodrigofranc — APPROVED** [review]

**Pull request merged**